Add experimental help parsers to project and package (#107)
PowerShell/Crescendo: Microsoft.PowerShell.Crescendo/src/experimental/HelpParsers/README.md

Eventually, we'll run out of sub-commands so the help will finally get parsed,
but if you notice the _order_ in which the parsed help objects are returned, they may not be the order you expect.

#### Parameter Recognition

Sometimes, the command argument would have an associated type for the parameter and sometime that would not be available.
This is why there are 2 patterns for a parameter:

```powershell
$parmPattern = "--(?<pname>[-\w]+)\s(?<ptype>\w+)\s+(?<phelp>.*)|--(?<pname>[-\w]+)\s+(?<phelp>.*)"
```

This pattern also ignores those patterns which are a single character.
I did this on purpose as PowerShell prefers parameters which are more descriptive.
I didn't create aliases out of those one character parameters, which is something that could be done in the future.
I also don't handle the cases where the one character parameter _follows_ the more verbose parameter.
In my inspection of the help, they all pretty much did something like:

```bash
	-q, --quiet     Do not be overly verbose
```

rather than:

```bash
	--quiet, -q     Do not be overly verbose
```

So my parameter recognizer expects it and doesn't capture the first part of the string.

> Again, this is probably pretty easy to support.

Reviewer: Expects what? I am not clear what "it" is referring to here. Also, I tested your pattern and it works the same for either case.

Additionally, sometimes the usage statement declares a parameter.
For example:

```powershell
PS> docker run --help|select -first 4

Usage:  docker run [OPTIONS] IMAGE [COMMAND] [ARG...]

Run a command in a new container
```

This indicates that there is a _positional_ parameter `IMAGE` as well as an optional parameter `COMMAND`.
The current parsers don't handle this case (they need to be added manually to the configuration).
This is something I would like to address in the future.

> **options _and_ arguments?** -

> Some of the tools have both named _and_ positional parameters, as well as options which apply to the _executable_ and the _command_.
> For example: `docker --debug image list --all` has a parameter `--debug` which applies to the `docker` executable,
> and `--all` which applies to the `image list` command.
> These two elements are an attempt to manage these various conditions.
> I'm not really satisfied with how I've done this, but haven't had time to tease apart the issues.
> I plan on getting back to it eventually.

Reviewer: ```suggestion
Some of the tools have both named _and_ positional parameters, as well as options that apply to the _executable_ and the _command_.
For example: `docker --debug image list --all` has a parameter `--debug` that applies to the `docker` executable,
and `--all` that applies to the `image list` command.
These two elements are an attempt to manage these various conditions.
I'm not really satisfied with how I've done this, but haven't had time to tease apart the issues.
I plan on getting back to it eventually.
```

#### Usage Recognition

This seemed to have the greatest variability across the tools.
Sometimes the usage statement has a clear delineation:

```powershell
PS> docker --help | sls usage:  -NoEmphasis

Usage:  docker [OPTIONS] COMMAND
```

but sometimes this (with an extra line):

```powershell
PS> kubectl --help
...
Usage:
  kubectl [flags] [options]
```

or this (with _multiple_ extra lines):

```powershell
PS> faas-cli --help|select -first 10

Manage your OpenFaaS functions from the command line

Usage:
  faas-cli [flags]
  faas-cli [command]
```

or this (next to Usage: and the following line):

```powershell
PS> netsh -?
Usage: C:\windows\system32\netsh.exe [-a AliasFile] [-c Context] [-r RemoteMachine] [-u [DomainName\]UserName] [-p Password | *]
             [Command | -f ScriptFile]
```

There was so much variability I sort of hand-waved at this.
I want to get back to this to see if there's a one-size-fits-all parser,
but the current examples are not very successful here.

#### The Packager

> Once the parsed help is separated into its components, we need a way to actually emit the Crescendo configuration.
> I chose to make the parser object model be able to render the text which comprises the Crescendo configuration.
> The parser command object has a method called `GetCrescendoCommand` which does the conversion from the parsers' object model to the Crescendo object model.

Reviewer: ```suggestion
Once the parsed help is separated into its components, we need a way to emit the Crescendo configuration.
I chose to make the parser object model be able to render the text that comprises the Crescendo configuration.
The parser command object has a method called `GetCrescendoCommand` that does the conversion from the parsers' object model to the Crescendo object model.
```

This makes the last mile of the problem the easiest, because the crescendo object _already_ knows how to build the JSON configuration.
Fortunately, we can rely on built-in features of .NET to do most of the work.
The only thing we need to do is to create an object which has the appearance of a configuration file:

```powershell
$h = [ordered]@{
    '$schema' = 'https://aka.ms/Crescendo/Schema.json'
    'Commands' = $commands
}
```

> After that, it's a simple matter of using the json serializer to construct the configuration.

Reviewer: ```suggestion
After that, it's a simple matter of using the JSON serializer to construct the configuration.
```

It _might_ be reasonable to emit the actual module,

> but my expectation is that the configuration will need some tweaking so emitting the `JSON` is the better approach.
> The parser takes the hashtable and some options and just calls the `JSON` serializer:

Reviewer: ```suggestion
but my expectation is that the configuration will need some tweaking so emitting the JSON is the better approach.
The parser takes the hashtable and some options and just calls the JSON serializer:
```

```powershell
System.Text.Json.JsonSerializer]::Serialize($h, $sOptions)
```

> **_And what about all that JSON_** -
As I was designing Crescendo, I wanted to be sure that the object model itself would be able to be expressed as configuration rather than code.
I knew that code would be needed for _some_ things, but I wanted to minimize how much code was required.

> JSON, with it's easy schema-tizing was a natural for this.

Reviewer: ```suggestion
JSON, with its easy schema validation was a natural for this.
```

> The schema can be annotated with tool-tips making the authoring process easier as well as supports mandatory elements, etc.

Reviewer: ```suggestion
The schema can be annotated with tool-tips making the authoring process easier, as well as support for mandatory elements, types, enums, etc.
```

As I was investigating this, I did consider using PowerShell native hash-tables, but they simply lack these extra features.

> Rather than spending my time in adding these features to PowerShell, _which I heartily recommend someone undertake_,

Reviewer: ```suggestion
Rather than spending my time adding these features to PowerShell, _which I heartily recommend someone undertake_,
```

I chose something that was ready to go.
I suppose I could have chosen `XML`...

## About these parsers

> These help parsers are all "work in progress", they're aren't finished or products but rather something that can be used as a starting point.

Reviewer: ```suggestion
These help parsers are all "works in progress". They aren't finished products but rather something that can be used as a starting point.
```
